Overhaul PostgreSQL packaging, native extensions, and NixOS support

thoughtpolice opened the pull request:
###### Motivation for this change

I'm brief on time now, but this isn't done yet. See #38616. In short, this overhauls the namespacing and extension management of PostgreSQL packages, greatly easing the ability to test, build, and package PostgreSQL extensions properly. It also overhauls the NixOS module to support the proposed API in #38616, making it far easier to add custom extensions to your server in a correct way, no matter what version you choose.

I've been running these changes with some light testing for a few days, but I'm not done yet. I'll be amending and adding more comments to the commits and maybe squash one or two. These changes otherwise seem solid and shouldn't break anything.

Note that the NixOS module _remains backwards compatible_. Nothing should break, and you're unlikely to even notice if you don't use extensions. I consider this a good thing and ideally supporting it to 18.09 should not be terrible. However, warnings will be thrown when using the old APIs, and assertions are thrown if both are used at once in incompatible ways. (In fact you can even use a mix of the old and new APIs, in theory -- e.g. use the new `.packages` attribute with the old `.extraPlugins`, instead of the new `.plugins`.) I'm open to dropping backwards compatibility if other people think we can give it up for 18.09, though.

Finally, this also updates Nixpkgs and NixOS to **use PostgreSQL 10 and libpq 10 by default**. Although this isn't necessary, it seems like a reasonable time to make the push, since 9.6 has been the default since 17.09. PostgreSQL 11 will ideally be out around September, near the 18.09 release, but it's unlikely we'll have time to fit it in by then (depending on how things go for us, and them.)

I have also stepped up and added myself as the maintainer of the PostgreSQL module, and will be finishing the manual section.

I suggest reviewing the commits sequentially, in order, rather than at once. They are all individually small, work, and should be much easier to follow.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Actually finish writing the PostgreSQL section of the manual.
- [x] Look into deprecation warnings for old top-level attrs; using `builtins.trace` is annoying because e.g. it triggers on `nix search -u ...`
- [x] Look into #22653 and #38469 as well -- these should be fixed while I'm at it, IMO.
- [x] ~Bikeshed the naming of the new options? `.packages` is really close to `.package`, but I'm not sure of a better name. On the other hand, I think `.plugin` is a good name vs `.extraPlugins`~
- [x] Migrate all tests and NixOS modules to new API

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [x] other Linux distributions (Fedora 25, sandboxing enabled)
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

thoughtpolice commented:
Please note that GitHub hates actually showing you the linear history of commits as it will appear in the history and instead always prefers showing things sorted-by-date, even when this is wrong. I rebased and re-arranged a few commits... Ugh.

thoughtpolice commented:
Note to future archaeologists who will later discover my corpse here: this was the magical incantation needed to make GitHub display commits in order, as it should, by resetting the commit/author date of every individual commit to be 2-3 seconds apart.

`for x in $(git log --format=oneline master..pgsql-fixes | tac | awk '{print $1}'); do git cherry-pick $x; git commit --amend --reset-author --no-edit; sleep 2; done`

Surely this will be the solution to an untold future terror and save the world, one day.

domenkozar commented:
@thoughtpolice can this go as-is and then we add further improvements?

thoughtpolice commented:
Yes, I've been running these changes in a stable manner for at least a few weeks on my development machine, though I was hoping @ocharles could have given a review -- but I noticed I forgot to add him!

thoughtpolice commented:
I was also hoping to add a `postgresqlWithPackages` function, so you could use postgres without NixOS, but with NIX_LIBDIR properly set up -- unfortunately this either requires some unsavory duplication of code between Nixpkgs and NixOS to retain backwards compatibility.

I believe I have a WIP patch for this, so I may add it as well.

danbst commented on nixos/modules/services/databases/postgresql.nix:
```diff
@@ -164,16 +201,46 @@ in
+      mkDefault (if versionAtLeast config.system.stateVersion "18.09" then pkgs.postgresql10Packages
+            else if versionAtLeast config.system.stateVersion "17.09" then pkgs.postgresql96Packages
+            else if versionAtLeast config.system.stateVersion "16.03" then pkgs.postgresql95Packages
+            else pkgs.postgresql94Packages);
```
Can this be made dependent on `cfg.package.version` instead of `config.system.stateVersion`?

thoughtpolice replied:
The intent is that e.g. if you were on NixOS 16.03, and you upgrade to NixOS 17.09, you should keep using PostgreSQL 9.5, because 9.6 is not on-disk compatible, and breaking your database across an upgrade isnt friendly. This is because when you installed NixOS, the (16.03) installer explicitly sets `stateVersion = "16.03"` in `configuration.nix` by default, and *that should not be changed* even beyond upgrades.

I might be misunderstanding what you mean, of course (a diff would be helpful if that's the case!)

danbst replied:
This isn't a direct problem by itself, but rather a simplification for those who use PG older than NixOS default. When you set `services.postgresql.package` to some old version, you now have to set `services.postgresql.packages` to an older pacakgeset too, which, I think, can be detected by NixOS automatically.

thoughtpolice commented:
@ocharles has given me his blessing/trust on this change in lieu of the fact he doesn't have much time to review it, but I'd love to start merging some of this in soon. I will post a message to the mailing list soon about this for any final objections.

samueldr commented:
This will need to be documented in the release notes (nixos/doc/manual/release-notes/rl-1809.xml) I presume.

As a user, I really much prefer this newly presented interface for PostgreSQL stuff. :+1:

thoughtpolice commented:
I'm still working on this, sorry for all the delays all. If it's any consolation, this has been running quite stable for me. :)

I've updated the release notes, issues like #22653 and #38469 should be fixed now (thanks to @basvandijk), and there is a beta of PostgreSQL 11 now available[1]. I'd like to actually get this done this week. Remaining things:

  - Finish manual (particularly backup/restore/upgrade cycle)
  - Finish migrating tests.

I expect this will probably take another few hours of focused work but I plan on landing this before the 18.09 branch.

---

[1] I am currently using this primarily to explore integration of LLVM/JIT support into the Postgres 11 betas; it's quite possible that I will *remove* the top-level attributes for postgres 11 before merging this, so people don't get any funny ideas, but in the mean time the support will be there and worked out, hopefully.

thoughtpolice commented:
Some more changes:

  - There's a new `.withPackages` combinator exposed from every package set, which allows you to drop into a Nix shell with postgres+plugins inside a `buildEnv`. This is mostly useful for doing upgrades or demonstrations, but in the future could be useful for other things, too (for example: packing PostgreSQL Docker images along with extensions, using Nix).
  - PGXS previously did not respect `PREFIX` during builds, meaning every Postgres extension had to have a custom, hackily written `installPhase`. I have fixed PGXS' to obey `PREFIX`, meaning `make install PREFIX=/nix/store/...` now works for the majority of extensions, so their installPhases can be removed. This is also necessary for JIT support, which requires shipping bitcode files during `installPhase`, and would be tedious to ship correctly.
  - JIT support for PostgreSQL 11 has been extended and should be "Ready to Ship" -- I fixed most glaring problems with this feature. Please see the corresponding `postgres11` commit messages for details. The net result though is that `postgresql11.out` itself will bloat from about ~130MB to ~400MB due to needing `libLLVM.so`, but everything else should be fine. (In particular, `libpq`-based clients will not bloat at all).
  - There's now a section for doing simple, `pg_dumpall`-based upgrades in the manual. It uses the `.withPackage` technique for showing a demonstration upgrade, since you want newer `pg_dumpall` tools.

thoughtpolice committed:
nixpkgs: deprecate the badly named 'postgresql100' attribute

See #38616. This is a minor issue, and for now we keep around the
old attribute name with a warning to tell people how to switch.

Signed-off-by: Austin Seipp <aseipp@pobox.com>

thoughtpolice committed:
nixpkgs: move postgresql patches into a common directory

This simply makes it easier to browse the PostgreSQL package directory. More to
come.

Signed-off-by: Austin Seipp <aseipp@pobox.com>

thoughtpolice committed:
nixpkgs: reorganize the postgresql extensions

This is a backwards compatible change; it mostly puts all the extensions
for postgresql in a common directory to keep them isolated.

It also moves a few things that /were not/ extensions out into other parts of
the filesystem namespace; namely the postgresql_jdbc and psqlodbc libraries
were moved under development/java-modules and development/libraries,
respectively. Because these libraries use the libpq postgresql client drivers,
they're less sensitive to underlying version changes anyway (since the protocol
is relatively stable).

No attributes were renamed or harmed in the creation of this patch.

Signed-off-by: Austin Seipp <aseipp@pobox.com>

thoughtpolice committed:
nixpkgs: tidy up pgjwt package

This makes pgjwt take a dummy 'postgresql' argument, which it does not *need*
in the buildInputs (it is purely a SQL extension with no C code). However, this
argument will be necessary for an upcoming change that will parameterize the
extensions over a particular PostgreSQL version.

It also does some tiny cleanup, setting a null build phase.

Signed-off-by: Austin Seipp <aseipp@pobox.com>

thoughtpolice committed:
nixpkgs: implement new postgresqlPackages infrastructure

This adds five new top-level attrsets:

    pkgs.postgresql93Packages
    pkgs.postgresql94Packages
    pkgs.postgresql95Packages
    pkgs.postgresql96Packages
    pkgs.postgresql10Packages

Each of these contains a set of PostgreSQL extensions, built against the
particular specified postgresql package. This makes it far easier for specific
PostgreSQL extensions to be built and used in NixOS modules, and provides an
easier way to test extension builds against multiple versions. For example, we
can now use either:

    pkgs.postgresql96Packages.timescaledb

or

    pkgs.postgresql10Packages.timescaledb

when building and specifying extensions.

By default, the prior top-level attribute names such as 'timescaledb' HAVE NOT
been removed. They will be deprecated in a future commit to encourage people to
move to explicitly-versioned attrsets.

Signed-off-by: Austin Seipp <aseipp@pobox.com>

thoughtpolice committed:
nixpkgs: migrate postgis to new postgresqlPackages infrastructure

Signed-off-by: Austin Seipp <aseipp@pobox.com>

thoughtpolice committed:
postgresqlPackages: add myself as a maintainer

Signed-off-by: Austin Seipp <aseipp@pobox.com>

thoughtpolice committed:
postgresqlPackages: add versionCheck passthru to extension builds for checking PostgreSQL compatibility

Not every PostgreSQL extension builds with every version we have available; for
instance, timescaledb is only available for PostgreSQL 9.6 and later.

To do this, extensions may use a 'passthru.versionCheck' attribute, which
should return 'true' if the extension can be built and used against the given
postgresql version. This avoids needless build failures and helps indicate
to the user that an extension is not available.

Unfortunately the PostgreSQL expression did not previously offer a 'version'
attribute to check against, so we must add it (as expected). This implies a
rebuild of the PostgreSQL derivations, unfortunately.

Signed-off-by: Austin Seipp <aseipp@pobox.com>

thoughtpolice committed:
postgresqlPackages.timescaledb: only works on 9.6+

Signed-off-by: Austin Seipp <aseipp@pobox.com>

thoughtpolice committed:
postgresqlPackages.pg_cron: only works on 9.5+

Signed-off-by: Austin Seipp <aseipp@pobox.com>

thoughtpolice committed:
postgresqlPackages.pg_topn: only works on 9.4+

Signed-off-by: Austin Seipp <aseipp@pobox.com>

thoughtpolice committed:
postgresqlPackages.pg_jobmon: init at 1.3.3

Signed-off-by: Austin Seipp <aseipp@pobox.com>

thoughtpolice committed:
postgresqlPackages.pg_partman: init at 3.1.2

Signed-off-by: Austin Seipp <aseipp@pobox.com>

thoughtpolice committed:
nixos/postgresql: add myself as the module maintainer

Signed-off-by: Austin Seipp <aseipp@pobox.com>

thoughtpolice committed:
nixos/postgresql: kill ugly hack, now that postgresql has 'version' attrs

Signed-off-by: Austin Seipp <aseipp@pobox.com>

thoughtpolice committed:
nixos/postgresql: add new nixos module API

Signed-off-by: Austin Seipp <aseipp@pobox.com>

thoughtpolice committed:
nixos/tests: migrate postgis test to new postgres API

Signed-off-by: Austin Seipp <aseipp@pobox.com>

thoughtpolice committed:
nixpkgs/nixos: switch to PostgreSQL 10 by default

Signed-off-by: Austin Seipp <aseipp@pobox.com>

thoughtpolice committed:
nixos/postgres: update manual

Signed-off-by: Austin Seipp <aseipp@pobox.com>

thoughtpolice committed:
nixpkgs: actually deprecate old postgresql extension names

Signed-off-by: Austin Seipp <aseipp@pobox.com>

thoughtpolice committed:
nixos/tests: migrate to new postgresql api

Signed-off-by: Austin Seipp <aseipp@pobox.com>

basvandijk committed:
postgresql: always create $out/bin (issue #22653).

This is needed because some PostgreSQL plugins don't have a bin
directory. If only these plugins are listed in cfg.extraPlugins buildEnv
will turn $out/bin into a symbolic link to ${pg}/bin. Lateron we try to
rm $out/bin/{pg_config,postgres,pg_ctl} which will then fail because
$out/bin will be read-only.

Signed-off-by: Austin Seipp <aseipp@pobox.com>

basvandijk committed:
postgresqlPackages.pg_journal: init at 0.2.0

Signed-off-by: Austin Seipp <aseipp@pobox.com>

thoughtpolice committed:
postgresqlPackages.tsearch_extras: 0.3 -> 0.4

Signed-off-by: Austin Seipp <aseipp@pobox.com>

thoughtpolice committed:
postgresqlPackages.pg_cron: 1.0.2 -> 1.1.2

Signed-off-by: Austin Seipp <aseipp@pobox.com>

thoughtpolice committed:
postgresqlPackages.pg_partman: 3.1.2 -> 3.2.1

Signed-off-by: Austin Seipp <aseipp@pobox.com>

thoughtpolice committed:
postgresqlPackages.pg_topn 2.0.2 -> 2.1.0

Signed-off-by: Austin Seipp <aseipp@pobox.com>

thoughtpolice committed:
postgresql: enable systemd service notification on 9.6+

This results in a marginal closure size increase of ~137mb -> ~141mb.
This only affects the 'out' output, not the 'lib' output, so
libpq-dependent expressions should not bloat as a result of this.

The type of the systemd service is changed to 'notify' on 9.6+, which
will help identify when Postgres is starting/shutting down.

Signed-off-by: Austin Seipp <aseipp@pobox.com>

thoughtpolice committed:
postgresqlPackages: simplify versionCheck comparisons

Extensions need to compare against the version of PostgreSQL they're
building against, so providing a partially applied function is easier
and less verbose for these cases.

Signed-off-by: Austin Seipp <aseipp@pobox.com>

thoughtpolice committed:
postgresqlPackages.pg_repack: (re)indent to 2 spaces

Signed-off-by: Austin Seipp <aseipp@pobox.com>

thoughtpolice committed:
nixos/manual: add 18.09 release notes for PostgreSQL

Signed-off-by: Austin Seipp <aseipp@pobox.com>

thoughtpolice committed:
postgresql: add postgresql 11.0 beta2 package

This also bars all extensions which currently do not build.

Signed-off-by: Austin Seipp <aseipp@pobox.com>

thoughtpolice committed:
postgresql11: preliminary JIT support

This adds preliminary support for JIT'ing SQL queries via LLVM to
PostgreSQL, enabled on versions 11+, on Linux. The default stdenv of
PostgreSQL itself and all exposed sub-packages are overridden to use a
particular major version of llvmPackages.stdenv so that 'clang' is
chosen as the default compiler. Currently, we use llvmPackages_6. To do
this, we expose another component from postgresqlXXPackages -- a new
.stdenv attribute, which exposes which stdenv PostgreSQL was built with.

PostgreSQL JIT support works by compiling all of the source code for the
system into LLVM bitcode, and shipping this as part of the binary
distribution (under the .lib output, in our case, $lib/lib/bitcode/). At
runtime, Postgres uses the LLVM API in order to load this bitcode and
JIT queries directly against the source code. This allows inlining
database code into queries, in particular inlining operator and
expression definitions for custom types, in any extensions. This feature
is enabled via the '--with-llvm' flag during configurePhase.

This feature is integrated with 'PGXS', the Makefile infrastructure for
writing and distributing Postgres extensions. If an extension is being
compiled against a version of PostgreSQL that has LLVM JIT support, then
the extension code is *also* compiled to bitcode and distributed
transparently in the binary output, alongside the extension shared
object. In order to ensure consistent bitcode is emitted, when Postgres
is compiled with LLVM support, it *must* be compiled with Clang (hence
the stdenv choice), and it also *must* embed the copy/path of Clang into
the resulting binaries, so PGXS can use the exact same compiler later on
during 3rd party extension builds. This design decision is likely
because on systems such as Debian, multiple versions of Clang can exist,
so any compiled Postgres code/extensions must use a consistent compiler
for all future builds.

However, this decision has extremely negative consequences for Nix-based
packages, because it inserts LLVM and Clang into the closure of the
PostgreSQL derivations as a hard runtime-dependency. This bloats the
closure size by over a gigabyte (~140MB -> 1.4GB), which is fairly
unwieldly and unlikely to be permissible by default. Currently this
bloat only applies to the .out (binary) outputs.

But this is made worse by the fact that the .lib output is also bloated
by having a hard runtime dependency on llvmPackages.llvm.lib. This is
because postgresql.lib now ships libllvmjit.so which talks to
libLLVM.so, but postgresql.lib also contains client libraries like
libpq.so. This effectively bloats every libpq client expression as well
by about 200MB.

Finally, because of the way PGXS's default installation logic works, it
wants to install binary artifacts into the postgresql lib/ directory,
which obviously isn't possible in the Nix store as it's read-only.
Hence, we create environments composed of all extensions outputs and
patch postgresql to load that. But this means the installPhase for every
extension is currently a custom hand-written script, and *that* means
every extension must now contain logic to install LLVM .bc files on top
of .sql and .so files.

Instead, we should probably patch PGXS to install to a proper external
directory so its install logic can take over and we can remove custom
installPhase scripts for most extensions. Postgres' LLVM support logic
in its configure script and PGXS code will need to be patched to remove
hard-coded references to clang-wrapper, since we always control the
exact version of clang used and can remove it as a run-time dependency.
Finally, bitcode should probably be moved to separate .bitcode
derivations for all server versions and extension outputs, so that libpq
clients aren't bloated by indirect dependencies on libLLVM.so (by way of
libllvmjit.so).

Oh, and PostGIS fails to build with JIT support/clang as the compiler,
for some reason.

As a result of these significant complications, this support is disabled
by default, and should only be considered supported for vanilla
PostgreSQL with no third-party extensions.

Darwin may also be supported in the future; it may even build, but can't
be tested.

Signed-off-by: Austin Seipp <aseipp@pobox.com>

thoughtpolice committed:
postgresql11: remove hard clang dependency for JIT support

When JIT support was enabled, PGXS embedded a copy of clang-wrapper's
path into its $out, resulting in a large increase in closure size due
to dragging the whole compiler along. A similar problem existed
previously by way of binutils 'ld', which was worked around. Perform
that same workaround for clang-wrapper, too.

With JIT support enabled, this reduces the closure size of
postgresql11.out from ~1.4GB to ~410MB.

Signed-off-by: Austin Seipp <aseipp@pobox.com>

thoughtpolice committed:
postgresql11: keep bitcode/llvmjit.so in $out, not $lib

With JIT support enabled, the Postgres build system keeps a copy of
bitcode for the server, and a new llvmjit.so extension, in the $lib
output. However, this bloats the closure of any libpq-dependent
libraries by requiring a hard dependency on libLLVM.so, as well as the
actual bitcode files (which aren't useful at all to clients).

Moving both of these into $out does not hurt binary output size (since
it already was dependent on $lib anyway), but does help all clients.

Signed-off-by: Austin Seipp <aseipp@pobox.com>

thoughtpolice committed:
postgresqlPackages: add new withPackages combinator

withPackages is like a local version of the services.postgresql.plugins
option: it allows you to build an environment containing PostgreSQL and
all the extensions you desire, by way of attrsets.

For example, the following command will drop you in a Nix shell
containing all the binaries from PostgreSQL 10, including the
TimescaleDB extension as a loadable library:

    $ nix-shell -p \
      'postgresql10Packages.withPackages (p: with p; [ timescaledb ])'

This is primarily useful for running demonstrations or local
experiments, but may also be useful for performing database upgrades (as
you may want pg_dumpall, etc from _newer_ PostgreSQL versions, to run
against your current server).

Signed-off-by: Austin Seipp <aseipp@pobox.com>

thoughtpolice committed:
postgresqlPackages.pg_similarity: remove needless dir

Signed-off-by: Austin Seipp <aseipp@pobox.com>

thoughtpolice committed:
postgresqlPackages: make PGXS install phase Nix-compatible

By fixing the (non-overrideable) PGXS installation directory variables
in the Postgres expression, any further extensions can be installed
using traditional pg_config-based PGXS Makefile build systems. This that
overriding 'installPhase' is no longer necessary; an ordinary PREFIX=...
will work, just like most Makefile build systems.

Aside from being cleaner, less-error prone, and easier to write, this is
also another piece necessary for useable JIT support.

Signed-off-by: Austin Seipp <aseipp@pobox.com>

thoughtpolice committed:
nixos/manual: finish section on PostgreSQL server upgrades

Signed-off-by: Austin Seipp <aseipp@pobox.com>

thoughtpolice committed:
postgresqlPackages: init citus 7.5.0

Signed-off-by: Austin Seipp <aseipp@pobox.com>

thoughtpolice committed:
postgresqlPackages.postgis: fix gdal against the same postgresql the extension is built against

GDAL's Postgres support had a weird cyclic dependency: GDAL depends on
Postgres (for possible spatial information querying for GDAL clients),
and PostGIS depends on GDAL. This sounds fine, except it has a dirty
follow up: GDAL is *only* built against the default version of
PostgreSQL. This means if you try to use a non-default version of
Postgres with PostGIS, it pulls in GDAL, which pulls in some other
version of Postgres. This results in weird situations like PostgreSQL
9.4 + PostGIS depending on PostgreSQL 9.6 in their closure.

In a previous experiment I tried making this optional, but there's a
better way: instead of making it optional, we simply let the default
'gdal' expression depend on 'postgresql' (whatever version the default
might be), and then we simply override the version of GDAL used by
postgis to be the *same* postgres that postgis is built against. This
effectively results in a private copy of gdal that only is used by the
postgis extension builds.

Signed-off-by: Austin Seipp <aseipp@pobox.com>

thoughtpolice committed:
postgresqlPackages: expose a set containing all postgresqlPackages sets, for testing

Occasionally, it may be useful to enumerate _all_ supported package
sets, for example, in the NixOS tests. This refactoring supports that
while keeping the level of redundancy the same, for the most part.

Signed-off-by: Austin Seipp <aseipp@pobox.com>

thoughtpolice committed:
nixos/postgresql-backup: migrate to new module API

Signed-off-by: Austin Seipp <aseipp@pobox.com>

thoughtpolice committed:
nixos/tests: touch up PostgreSQL tests, migrate to new module API

Signed-off-by: Austin Seipp <aseipp@pobox.com>

thoughtpolice committed:
nixos/tests: wibble for postgis tests [NFC]

Signed-off-by: Austin Seipp <aseipp@pobox.com>

thoughtpolice committed:
nixos/tests: touch up pgjwt tests

Signed-off-by: Austin Seipp <aseipp@pobox.com>

thoughtpolice committed:
nixos/tests: migrate trac to new postgresql API

This test doesn't work, but it suppresses a needless warning, anyway.

Signed-off-by: Austin Seipp <aseipp@pobox.com>

thoughtpolice committed:
nixos/virtualisation: fix example

Signed-off-by: Austin Seipp <aseipp@pobox.com>

thoughtpolice committed:
nixos/gitea: update to new PostgreSQL interface

Signed-off-by: Austin Seipp <aseipp@pobox.com>

thoughtpolice committed:
nixos/gitlab: use new postgres module API

Signed-off-by: Austin Seipp <aseipp@pobox.com>

thoughtpolice committed:
nixos/matrix-synapse: use new postgres module API

Signed-off-by: Austin Seipp <aseipp@pobox.com>

thoughtpolice committed:
postgresqlPackages.postgis: enable PostgreSQL 11 JIT support

This was a bug that was fixed upstream; the applied patch is slightly
different but with the same result. Should be removed when a new version
of PostGIS is out.

Signed-off-by: Austin Seipp <aseipp@pobox.com>

basvandijk reviewed (changes requested):
Impressive work! 18.09 is getting better and better.

thoughtpolice committed:
postgresqlPackages: remove outdated comments

Signed-off-by: Austin Seipp <aseipp@pobox.com>

thoughtpolice committed:
nixos/manual: add more postgresql modules to 18.09 release notes

Signed-off-by: Austin Seipp <aseipp@pobox.com>

basvandijk commented:
@thoughtpolice I added extra tests for some extensions in:

https://github.com/thoughtpolice/nixpkgs/compare/pgsql-fixes...LumiGuide:pgsql-fixes-extra-tests

I leave it up to you to decide to merge this into this your PR.

I made it so the extra tests are easily disableable which is handy if you want to debug the tests and don't want the extra complexity of certain extensions.

@nbp in the process I discovered that merging options like [`services.postgresql.plugins`](https://github.com/NixOS/nixpkgs/blob/7e937697dcad230880d541c339716c9f6f06fa76/nixos/modules/services/databases/postgresql.nix#L143) produce evaluation errors. These options should be assigned a function that takes an attribute set and returns a list with a selection of the values in the input set. Merging these options should create a new function that applies the functions of all the options to the input set and concatenate the resulting lists.

For this [I created the `selectorFunction` type](https://github.com/thoughtpolice/nixpkgs/commit/fec24c27ddf7d34b2252d2068349fdb44995b00c). It's currently defined in the postgresql module but there are at least 6 other places in nixpkgs that use these type of options:

* [`services.gitit.extraPackages`](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/misc/gitit.nix#L44)
* [`services.ihaskell.extraPackages`](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/misc/ihaskell.nix#L22)
* [`services.octoprint.plugins`](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/misc/octoprint.nix#L68)
* [`services.xserver.windowManager.xmonad.extraPackages`](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/x11/window-managers/xmonad.nix#L30)
* [`services.xserver.windowManager.exwm.extraPackages`](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/x11/window-managers/exwm.nix#L27)
* [`services.hoogle.packages`](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/development/hoogle.nix#L27)

What do you think about moving the `selectorFunction` type to `lib/types.nix` and patching these module to use it?

basvandijk committed:
nixos/postgresql: give the plugins option the selectorFunction type

This allows to merge configurations that both define plugins as in:

  mkMerge [
    { services.postgresql.plugins = p: [ p.pg_journal ]; }
    { services.postgresql.plugins = p: [ p.postgis ]; }
  ]

selectorFunction should be moved to lib/types.nix since options like
"plugins" occur in at least 6 other places in NixOS.

Signed-off-by: Austin Seipp <aseipp@pobox.com>

thoughtpolice committed:
nixos/postgresql: give the postgres user a homedir and shell

For certain SQL operations such as

    COPY (...) [TO|FROM] PROGRAM '...' WITH (FORMAT CSV)

the designated PostgreSQL server user must be able to execute the
specified program and pipe the data into it; however, Postgres executes
these commands *under a shell* in order to support features such as
traditional piping, for commands like split or gzip. With no homedir and
no shell assigned, Postgres fails to execute the shell, resulting in a
hang that then later results in the query failing. Depending on the
program and query executed, this may happen at seemingly random times
due to buffering/piping issues.

While it isn't traditional to assign a shell to the database user (at
least on NixOS), doing so restores this functionality, which is
extremely useful for e.g. bulk ETL pipelines that copy to/from external
data files.

Signed-off-by: Austin Seipp <aseipp@pobox.com>

thoughtpolice committed:
timescaledb-parallel-copy: init at 2018-05-14

Signed-off-by: Austin Seipp <aseipp@pobox.com>

thoughtpolice committed:
postgresqlPackages: add .man to all withPackage/NixOS configurations

Otherwise, features like 'man psql' don't work, which is weird for a
machine that has PostgreSQL running as a service.

Signed-off-by: Austin Seipp <aseipp@pobox.com>

thoughtpolice committed:
postgresqlPackages: enable LLVM JIT by default on PostgreSQL 11+

This turns on JIT support by default for PostgreSQL 11-compatible
packages, now that most of the problems are fixed:

  1) Closure sizes don't include Clang, only LLVM,
  2) libpq clients don't suffer bloat from libLLVM.so dependencies,
  3) PGXS support is improved, so hand-written installPhases are
     mostly not needed, making it much more viable to ship bitcode
     for 3rd party extensions.

As a side note, this change also should not impact the nixos-small
channel builds very much (I hope), because those already have LLVM 6.0
in their closure, by way of QEMU/mesa packages.

Signed-off-by: Austin Seipp <aseipp@pobox.com>

basvandijk commented:
What prevents us from cherry-picking all of this on `release-18.09` before the release publication?

basvandijk commented:
@thoughtpolice I merged `master` into `pgsql-fixes` in [LumiGuide:pgsql-fixes-master-merge](https://github.com/NixOS/nixpkgs/compare/master...LumiGuide:pgsql-fixes-master-merge) and fixed the conflicts. `nix-build nixos/tests/postgresql.nix` succeeds.

However, I realize it might actually be better to rebase `pgsql-fixes` on `master` to make it easier to cherry-pick the commits on `release-18.09` later on if we decide that we want to do that. I'll do that later this evening.

basvandijk committed:
postgresql: add doc section id module-services-postgres-new-extensions

basvandijk commented:
I rebased on `master` to fix the eval errors.

@GrahamcOfBorg test postgresql

GrahamcOfBorg commented:
<!--REQUEST_ID=af9b4fbf-8761-4ba8-8c33-eb0d58fbf91f-->
Success on x86_64-linux

Attempted: tests.postgresql

No partial log is available.

GrahamcOfBorg commented:
<!--REQUEST_ID=af9b4fbf-8761-4ba8-8c33-eb0d58fbf91f-->
Success on aarch64-linux

Attempted: tests.postgresql

No partial log is available.

basvandijk commented:
Cherry-pick of this PR on `release-18.09`: https://github.com/NixOS/nixpkgs/compare/release-18.09...LumiGuide:pgsql-fixes-release-18.09. 

The postgresql tests pass. The manual incl. release notes build successfully and look correct.

basvandijk commented:
I missed that we still have an unchecked TODO item: "Migrate all tests and NixOS modules to new API". Indeed modules like [`hydron`](https://github.com/NixOS/nixpkgs/blob/9a88df84997d0c6ba57132156ac5f22496535c1d/nixos/modules/services/web-servers/hydron.nix#L108) haven't been migrated yet.

Every module that revers to `config.services.postgresql.package` needs to replace that with:
```nix
if (config.services.postgresql.package != null) 
then config.services.postgresql.package 
else config.services.postgresql.packages.postgresql
```
This is quite cumbersome and error-prone to repeat everywhere.

@thoughtpolice may I propose that we turn the local variable [`postgresqlPackage`](https://github.com/NixOS/nixpkgs/blob/9a88df84997d0c6ba57132156ac5f22496535c1d/nixos/modules/services/databases/postgresql.nix#L18) into a [`readOnly`](https://github.com/NixOS/nixpkgs/blob/9a88df84997d0c6ba57132156ac5f22496535c1d/lib/options.nix#L22) option of the `postgresql` module? Then other modules can refer to that option abstracting over the if..then..else.

basvandijk committed:
postgresql: introduce the deprecated, readOnly and invisible postgresqlPackage option

This option is used in modules that need to refer to the selected
PostgreSQL package. Without this option these modules need to
duplicate some conditional logic which is error-prone.

This option should be removed when the deprecated
config.services.postgresql.package is removed.

basvandijk commented:
@thoughtpolice I've now introduced the read-only and invisible `config.services.postgresql.postgresqlPackage` option in the last commit of https://github.com/NixOS/nixpkgs/compare/master...LumiGuide:pgsql-fixes-postgresqlPackage (https://github.com/NixOS/nixpkgs/commit/a1d18b73a3388f6a2f0275adb8bf337f1aa2336d). That commit also migrates all remaining modules to the new scheme. What do you say? Merge it here?

vcunat commented:
It's better, unless there's something that's urgent (e.g. security fixes).  You just merge to the staging branch.  

Everything is manual.  Someone chooses some version of staging to stabilize &ndash; merges it to `staging-next` to stop inflow of other large changes, then repeatedly regressions are fixed (if any) and then it's merged to master.  We recently added that intermediate step https://github.com/NixOS/rfcs/pull/26 (not merged yet :)

basvandijk commented:
@vcunat thanks for the excellent explanation and RFC. Then I think I will be merging this into `staging` soon.

@vcunat as 18.09 release manager, is there a chance that [this can still be cherry-picked](https://github.com/NixOS/nixpkgs/compare/release-18.09...LumiGuide:pgsql-fixes-release-18.09) in `release-18.09` once this is merged in `master`? Or is that out of the question?

@GrahamcOfBorg test postgresql.postgresql10  postgresql.postgresql11  postgresql.postgresql93  postgresql.postgresql94  postgresql.postgresql95  postgresql.postgresql96

GrahamcOfBorg commented:
<!--REQUEST_ID=7a010225-f377-44ab-8114-1bc975771310-->
Success on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.38698&attempt_id=5d7dea72-e361-4359-a5e2-3340cca49f0a)

Attempted: tests.postgresql.postgresql10, tests.postgresql.postgresql11, tests.postgresql.postgresql93, tests.postgresql.postgresql94, tests.postgresql.postgresql95, tests.postgresql.postgresql96

<details><summary>Partial log (click to expand)</summary><p>

```
syncing
test script finished in 65.41s
vde_switch: EOF on stdin, cleaning up and exiting
cleaning up
/nix/store/mk98qvgs2w57i2fxc2g1c3amydp9ky9g-vm-test-run-postgresql10
/nix/store/j80bwkkqibwhrvy6rprlvpm1114f12xf-vm-test-run-postgresql11
/nix/store/l7fr93vxk247903k5iq04pq21s1zvn2x-vm-test-run-postgresql93
/nix/store/mh9p9njljlkr31saqxip7crpz3y0dn9w-vm-test-run-postgresql94
/nix/store/dp3n1hhydxxnazjwxph4qws0zyv8k59n-vm-test-run-postgresql95
/nix/store/zhi50awz9lhiq057kpvrfh6k08yniy71-vm-test-run-postgresql96
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=7a010225-f377-44ab-8114-1bc975771310-->
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.38698&attempt_id=63103663-cfd5-46ef-8846-f90ff8ab5250)

Attempted: tests.postgresql.postgresql10, tests.postgresql.postgresql11, tests.postgresql.postgresql93, tests.postgresql.postgresql94, tests.postgresql.postgresql95, tests.postgresql.postgresql96

<details><summary>Partial log (click to expand)</summary><p>

```
syncing
test script finished in 121.01s
vde_switch: EOF on stdin, cleaning up and exiting
cleaning up
/nix/store/r6shs002pfg4fvw0avs2znz2fwky2v3f-vm-test-run-postgresql10
/nix/store/l41aj23bdi3wkya4wi1rvvznj3zrjvwr-vm-test-run-postgresql11
/nix/store/4r0nwlm32cjz2nwr3afs25dx2niyyhv0-vm-test-run-postgresql93
/nix/store/mpp9kzkaaffl33v450yf79ybjg64hwsf-vm-test-run-postgresql94
/nix/store/9qxywinb9kn5wnh7dh3mpnsqfzxkh26q-vm-test-run-postgresql95
/nix/store/wxxrnxsvrx8p8ghw2hjki0gwz2mh4hsm-vm-test-run-postgresql96
```
</p></details>

thoughtpolice commented:
Sorry about slacking on this folks, I was sick during the beginning of September which meant I missed the fork deadline. :( I'm willing to shepard these changes in still, providing we have enough time.

I'll review the rest of the info in this thread shortly.

basvandijk commented:
@thoughtpolice cool to have you back. Let me know how I can help.

thoughtpolice commented:
So I think the immediate thing we should do is merge this to `staging` ASAP, in order to go ahead and get this in the pipeline to hit `master`. What I will probably do is rebase things again in order to get them in a clean state on top of `staging`, since it's still active right now, and it will trickle in through `staging-next`. I think most of the downstream consumers (the 1,200 packages) are mostly just due to PostgreSQL 10 becoming the default, requiring all `libpq` clients to be rebuilt.

Doing this I think will mostly just require fixing up the remaining tests and hitting The Big Button. (Unfortunately there was also a regression in the PostgreSQL 11 JIT support, meaning the closure size is bloated again, but I think I can fix it easily.)

The NixOS 18.09 question is trickier. I have asked @samueldr about an exception on IRC, and I would also like @vcunat's opinion, so I await their judgement. @basvandijk, it looks like you're already doing this in your LumiGuide fork[1] -- so if we get an exception, I think we can use that as the basis for a cherry pick into `staging-18.09` nicely as well, which will hopefully get brought in one final time before the release.

In the event this is barred, I'd suggest we can keep maintaining an independent fork that we merge with `release-18.09` regularly, for people who want it, although it would be unfortunate (but understandable).

Realistically, my biggest concern about merging this to 18.09 is if the API needs to change again for 19.03, due to feedback from users. That would cause some annoying churn between two releases. *But*, I think even if I had merged this a month ago, we'd still be in the same place -- the vast majority of users would not be updating or testing this new API until after the release _anyway_. Meaning the end result would have been the same: we probably would only have reliable feedback well after "The point of no return". I think that's probably just part of the game...

Alternatively, if this hit 18.09 --  we could take any feedback from this new API and roll it into 19.09, in order to ensure we have adequate time for cleanup of prior code, and any future amendments, leaving the gap of 19.03 for one release. This would also be livable, in my opinion.

@vcunat, @samueldr -- please let us know what you think, and @basvandijk and I can certainly handle it.

---

[1] Just curious, are you already using this work at LumiGuide? I've been using it successfully for several months on my server, so I'm curious.

basvandijk commented:
@thoughtpolice sounds like a plan.

> Just curious, are you already using this work at LumiGuide?

No, not yet but today I started the migration to 18.09 so hopefully I will be able to use it soon.

@jchildren I see you're giving a talk at [NixCon 2018](https://nixcon2018.org/#program) about building and testing PostgreSQL extensions with Nix. So you're probably interested in this work as well.

thoughtpolice commented:
Wait, @basvandijk -- how did you push to *my* fork exactly? I see this change on https://github.com/thoughtpolice/nixpkgs/tree/pgsql-fixes, but how can you push there? I don't have any collaborators, so this seems really weird!

samueldr commented:
> The NixOS 18.09 question is trickier. I have asked @samueldr about an exception on IRC, and I would also like @vcunat's opinion, so I await their judgement. 

For the record [the conversation we had](https://logs.nix.samueldr.com/nixos-dev/2018-09-11#1536690638-1536691453;).

I'm not too hot on changing the semantics of stateVersion == 18.09 so close to the release, but not totally entirely against it. Though, combined with the idea that this is a huge change post-freeze, it's a bit spooky IMHO. My main thinking was seeing how anybody else concerned would react to the backport news, and seeing only agreement I would go for it, *since we're still at the beta stage*. Though, the roadmap makes me think it'll be late beta, and possibly hard to go for it.

Anyway, let's see other opinions on backporting post feature-freeze.

danbst commented:
I guess those who rely on Postgresql will appreciate the fix of packaging situation, even if it requires postfixups in release branch.

basvandijk commented:
Since this PR also touches the vital hydra module:
@GrahamcOfBorg test hydra

GrahamcOfBorg commented:
<!--REQUEST_ID=718d3fe7-cef8-41ce-982e-233c07e7e470-->
No attempt on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.38698&attempt_id=a7b3c84b-1cfa-471b-ac8b-784af304a11f)

The following builds were skipped because they don't evaluate on aarch64-linux: tests.hydra

<details><summary>Partial log (click to expand)</summary><p>

```

a) For `nixos-rebuild` you can set
  { nixpkgs.config.allowUnsupportedSystem = true; }
in configuration.nix to override this.

b) For `nix-env`, `nix-build`, `nix-shell` or any other Nix command you can add
  { allowUnsupportedSystem = true; }
to ~/.config/nixpkgs/config.nix.


```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=718d3fe7-cef8-41ce-982e-233c07e7e470-->
Failure on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.38698&attempt_id=1594a56d-50eb-4fe6-8a30-38359f7d7c30)

Attempted: tests.hydra

<details><summary>Partial log (click to expand)</summary><p>

```
Sep 12 07:15:29 machine systemd[1]: Failed to start PostgreSQL Server.
Hint: Some lines were ellipsized, use -l to show in full.
error: command `systemctl status postgresql.service' did not succeed (exit code 3)
command `systemctl status postgresql.service' did not succeed (exit code 3)
cleaning up
killing machine (pid 600)
vde_switch: EOF on stdin, cleaning up and exiting
vde_switch: Could not remove ctl dir '/build/vde1.ctl': Directory not empty
builder for '/nix/store/nr0dlxd6s49a4w27vpw2an79gh0xlb70-vm-test-run-hydra-init-localdb.drv' failed with exit code 255
error: build of '/nix/store/nr0dlxd6s49a4w27vpw2an79gh0xlb70-vm-test-run-hydra-init-localdb.drv' failed
```
</p></details>

basvandijk commented:
The hydra test failure is due to the postgresql systemd service taking more than 2m to start up causing a timeout. Probably because ofborg is under heavy load. The test succeeds on my system.

basvandijk commented:
@thoughtpolice is this already merged with `staging`? (I don't have my laptop with me so I it's hard to check myself).

thoughtpolice commented:
No, I have some WIP branch to merge/rebase it on staging  myself; meant to get to it this weekend but was distracted the whole time. Should come later tonight.

domenkozar commented:
I've pushed 29660a208552a1e32f872333d6eb52e13226effa (and to 18.09) using same patch as in this PR, as it was really annoying me :)

basvandijk commented:
@thoughtpolice to fix the merge conflicts I've rebased `pgsql-fixes` on `master` again at my [`pgsql-fixes-master-rebase-2018-09-28`](https://github.com/NixOS/nixpkgs/compare/master...LumiGuide:pgsql-fixes-master-rebase-2018-09-28) branch. Shall I force-push that to this PR or do you want to take the lead with this?  Note that `staging` merges without conflicts into `pgsql-fixes-master-rebase-2018-09-28`.

danbst commented:
API to override postgresqlPackages isn't exposed.

I can make an override for `postgresql`:
```
       pg11 = (import lumiPkgs { }).postgresql11Packages.postgresql.overrideAttrs (super: rec {
         name = "postgresql-${version}";
         version = "11beta4";
         src = pkgs.fetchurl {
           url = "mirror://postgresql/source/v${version}/${name}.tar.bz2";
           sha256 = "0qxlfh1a7bhhamrbs3msk71pny7jxx0c0fs26zlmp7jjn138zqi1";
         };
       });
```
but can't convert it into `postgresql11beta4Packages` - `makePackageSet` isn't exposed. And extensions won't be built against my patched postgresql.

danbst commented:
They way, how extensions are filtered by PG version, isn't user friendly. For example, `pg_repack` is said to be compatible only with PG < 11, so it is absent from `postgresql11Packages`. This results into `attribute pg_repack not found` error, which wasn't obvious "why" (I know it should be there, because I've packaged it!).

Maybe show some warning "pg_repack extension is available only for PG < 11" too?

danbst commented:
@thoughtpolice @basvandijk PG11 is released, let's merge this?

thoughtpolice commented:
Unfortunately #48664 and related changes have caused a bit of a rebasing nightmare in the mean time due to attribute-related changes, so there's a cascade of fixes I need to do. This will also complicate the merge to staging, I think. I plan on spending the rest of the day trying to work this out but it'll probably take some time.

In the mean time I'm going to add myself as the owner of Postgres-related changes under Nixpkgs to make myself a blocking reviewer, so this doesn't happen in the future and I have time to review the changes people make, e.g. this PR already has PG 11 so the prior one was a bit redundant in the long run. (Not that anyone could be blamed about this, really, except for myself perhaps for not doing this sooner, and not keeping up with the changes people do submit!)

thoughtpolice commented:
Update: I made a lot of good progress yesterday but I basically slowly rewrote and compressed several of the patches from the original branch to be fewer in number, and easier to grok. I hope later today I'll be ready to overwrite this branch with a new version composed on `staging` instead and just slap the merge button.

thoughtpolice commented:
4569ee7d74d505f0a9b3fb8754951393e8be7e6f also fixes a recent-ish plv8 build failure.

domenkozar commented:
I think this PR is a good example why we encourage contributions to be split up into small pieces.

nixpkgs moves *very* fast and PRs that hang around half a year, accumulating changes across 50+ files are too big pain to review and merge in timely manner (given that we work async).

thoughtpolice commented:
I've mostly been gone for vacation and thanksgiving, FWIW. I just didn't get around to posting my staging delta, which is now substantially smaller than this PR (in number of commits and changes.) In fact, prior to my leave, I *was* piece-wise upstreaming several pieces directly into staging -- hence why I was doing several regular merges (master -> staging) a few weeks ago, to prep things (to make sure `master` wouldn't needlessly conflict, later on). So my changes have gotten smaller and more pointed, I believe.

The real issue is that multiple changes here cause mass-rebuilds and reordering them or splitting them up doesn't really change that fact, so you might as well change a bunch of it at once, if you ask me. From my count, my new branch has roughly 5 major commits that will rebuild the vast majority of all PostgreSQL builds and their downstream consumers. They each imply large, delicate changes so keeping them distinct is proper. I consider Hydra and Borg somewhat sacred resources so I try not to stress them too hard without a good reason. (It also improves my cycle time to do this anyway: my local machine is extremely fast and can build hundreds of complex expressions on the order of minutes, giving me much better turnaround than borg, at least until I hit the 'long tail' of every-downstream consumer, where I just can't have enough cores.) This isn't really a case of me being "right" though, it's just a justification for why I've acted so.

That said I understand the frustration; the reality is that my time is simply much slimmer these days (for a number of reasons, ranging from being busy to personal health) and I'm not actively paid to do maintenance like this, but at the time I needed it, I really wanted it. It probably does not help that I'm a fairly established maintainer -- I can willingly deal with occasional rebases and whatnot for my personal work and I have a good idea of what to do if things go wrong. But many ordinary users (as well as other devs) who want these changes now probably don't want to do that, understandably.

FWIW, my current delta looks like this for `staging`:

```
93e05631b04390a94ecbd7ebd44d8e5c93b81d7a postgresqlPackages.postgis: remove outdated comment
230049105532faf9e16954fde92b8859ced5055a postgresqlPackages: add new withPackages combinator, fix man page
68216f892e08d11af411b5907a95c813e494135c postgresql_11: use llvmPackages_7
eaac5719c672a8743141692e0f14945202108504 postgresql11: implement and enable JIT support
80e9a93579adabfbc63cb1310d4759fd42f1ccfc postgresqlPackages: make PGXS install phase Nix-compatible
6590a941ea9190af871ee51d192e86aec6bdf226 postgresql: always create $out/bin
bc6996beb431b4b925bd4cbad25145c85238000c nixos/tests: migrate to new postgresql api
9802c1eb06684d0ff3a16821bc82e513fc04dc25 nixos/tests: migrate postgis test to new postgres API
2dae770378b17efa617554f9a750b6a14a7694fb nixos/postgresql: give the plugins option the selectorFunction type
f9c33be8329f345f3ceb5dce322e83a663094474 nixos/postgresql: add new nixos module API
f195107caad0a26c533a715e45a249a375ee3038 postgresql: enable systemd service notification on 9.6+
9b86b39cc8792e03d3779992484d733da47663a3 postgresqlPackages: add versionCheck passthru for checking server compatibility
74ef5728a645fdda3b4f3b09e212d53c6e144593 nixpkgs: migrate postgis to new postgresqlPackages infrastructure
d0a41fde11b8706268a5f6692a13efda158f5952 nixpkgs: implement new postgresqlPackages infrastructure
c876b993145acdcd7e605faf9744690292b3a910 nixpkgs: reorganize the postgresql extensions
3afeb2d7895512754f68ef6335846346ffa7fd60 nixpkgs: move postgresql patches into a common directory
```

I can of course go ahead and merge *that* to staging. In fact it should work quite fine, I think. So perhaps I should finally start cycling `borg` once I get these changes posted. (Some of it is arguably out of scope, such as the LLVM work; this could also be removed, I suppose, but again it's going to cause a big rebuild anyway.)

This is quite minimal and isolated in my opinion; close to what you'd need to hit all the high points. The remaining points are:

  - Test migration isn't actually done
  - Manual needs to be rewritten (I had most of this but the merge conflicts were too great to work out, due to a pass of automatic formatting, so this is "done" but I have to reformat, prepare it, etc).

Other than that I do think it's almost ready. Realistically though I'm just very busy and testing all these combinations takes a bit of time with each iteration and packaging them up nicely; I think fundamentally the scope of the changes is a little large no matter how you slice it.

danbst commented:
@thoughtpolice probably we can split this PR somehow, to speedup merge? Don't want to miss pg11 in 19.03

danbst commented:
I've started split. My plan:
- [x] `postgresqlPackages`, and make them extensible! https://github.com/NixOS/nixpkgs/pull/54319
- [ ] fix stuff, so all should work with pg11
- [ ] change default to pg11. This is minimal plan for 19.03. Eveything else can go step by step
- [ ] enable JIT in pg11
- [ ] make aliases low-prio, so they don't appear in nix search. Instead, force postgresqlPackages to be present there
- [ ] rething NixOS module options (in regard to plugins)
- [ ] everything else

thoughtpolice commented:
I had actually started up on this again last week, but I've basically been in-and-out of the house quite a lot. I also split up some things that I think can go into staging quicker now, so I'll try to take a look at #54319 when I get a chance and we'll see if we can meet in the middle.

danbst commented:
@thoughtpolice this PR is heavily outdated now, I'll close it. Most of this stuff is already ported. The only non-ported things (AFAIK):
- JIT support
- PG version constraints for extensions

Feel free to open a PRs for that.
